File: backend/app/src/main/kotlin/net/animeta/backend/controller/v3/AuthenticateController.kt
package net.animeta.backend.controller.v3

import net.animeta.backend.exception.ApiException
import net.animeta.backend.service.AuthService
import org.springframework.http.HttpStatus
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class AuthenticateController(
    private val authService: AuthService
) {
    data class Params(
        val username: String,
        val password: CharArray,
        val persistent: Boolean
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as Params

            if (username != other.username) return false
            if (!password.contentEquals(other.password)) return false
            if (persistent != other.persistent) return false

            return true
        }

        override fun hashCode(): Int {
            var result = username.hashCode()
            result = 31 * result + password.contentHashCode()
            result = 31 * result + persistent.hashCode()
            return result
        }
    }

    data class Result(
        val sessionKey: String,
        val expiryMs: Long?
    )

    @PostMapping("/v3/Authenticate")
    @Transactional
    fun handle(@RequestBody params: Params): Result {
        val user = authService.authenticate(params.username, params.password)
        if (user == null || !user.active) {
            throw ApiException("없는 아이디거나 암호가 틀렸습니다.", HttpStatus.UNAUTHORIZED)
        }
        val session = authService.createSession(user, params.persistent)
        return Result(
            sessionKey = session.sessionKey,
            expiryMs = session.expiry?.toMillis()
        )
    }
}

File: backend-ts/src/controllers/authenticate.controller.ts
import { Body, Controller, HttpCode, HttpStatus, Post } from "@nestjs/common";
import { AuthService } from "src/services/auth.service";
import { ApiException } from "./exceptions";

type Params = {
  username: string;
  password: string;
  persistent: boolean;
};

type Result = {
  sessionKey: string;
  expiryMs: number | null;
};

@Controller('/api/v4/Authenticate')
export class AuthenticateController {
  constructor(
    private authService: AuthService,
  ) {}

  @Post()
  @HttpCode(200)
  async handle(@Body() params: Params): Promise<Result> {
    const user = await this.authService.authenticate(params.username, params.password)
    if (!user || !user.is_active)
      throw new ApiException('없는 아이디거나 암호가 틀렸습니다.', HttpStatus.UNAUTHORIZED)
    const session = this.authService.createSession(user, params.persistent)
    return {
      sessionKey: session.sessionKey,
      expiryMs: session.expiry?.total({unit: 'millisecond'}) ?? null,
    }
  }
}
